lib/main.rb

```diff
@@ -0,0 +1,168 @@
+      true
+    elsif given
+      if given.is_a? Regexp
+        my_each { |x| return false unless x =~ given }
+      elsif given.is_a? Class
+        my_each { |x| return false unless x.is_a? given }
+      else
+        my_each { |x| return false unless x == given }
+      end
+    else
+      my_each { |x| return false unless x }
+    end
+    true
+  end
+
+  def my_any?(given = nil)
+    if block_given?
+      my_each { |x| return true if yield(x) }
+      false
+    elsif given
+      if given.is_a? Regexp
+        my_each { |x| return true if x =~ given }
+      elsif given.is_a? Class
+        my_each { |x| return true if x.is_a? given }
+      else
+        my_each { |x| return true if x == given }
+      end
+    else
+      my_each { |x| return true if x }
+    end
+    false
+  end
+
+  def my_none?(given = nil)
+    if block_given?
+      my_each { |x| return false if yield(x) }
+      true
+    elsif given
+      if given.is_a? Regexp
+        my_each { |x| return false if x =~ given }
+      elsif given.is_a? Class
+        my_each { |x| return false if x.is_a? given }
+      else
+        my_each { |x| return false if x == given }
+      end
+    else
+      my_each { |x| return false if x }
+    end
+    true
+  end
+
+  def my_count(number = nil)
+    total = 0
+    if block_given?
+      my_each { |x| total += 1 if yield(x) }
+    elsif number.nil?
+      my_each do |_|
+        total += 1
+      end
+    else
+      my_each { |x| total += 1 if number == x }
+    end
+    total
+  end
+
+  def my_map(proc = nil)
+    new_arr = []
+    return enum_for(:my_map) unless block_given?
+
+    if proc
+      my_each do |element|
+        new_arr << proc.call(element)
+      end
+    elsif proc.nil? && block_given?
+      my_each do |element|
+        new_arr << yield(element)
+      end
+    end
+    new_arr
+  end
+
+  def my_inject(given = self[0], symbol = nil)
+    if block_given?
+      given ||= 0
+      index = 1
+      while index < length
+        given = yield(given, self[index])
+        index += 1
+      end
+      given
+    elsif (given.is_a? Symbol) || (symbol.is_a? Symbol)
+      if given.is_a? Symbol
+        case given
+        when :+
+          total = 0
+          my_each { |x| total += x }
+        when :-
+          total = self[0]
+          self[1..-1].my_each { |x| total -= x }
+        when :*
+          total = self[0]
+          self[1..-1].my_each { |x| total *= x }
+        when :/
+          total = self[0]
+          self[1..-1].my_each { |x| total /= x }
+        end
+        total
+      elsif given.is_a? Numeric
+        case symbol
+        when :+
+          to_a.my_each { |x| given += x }
+        when :-
+          to_a.my_each { |x| given -= x }
+        when :*
+          to_a.my_each { |x| given *= x }
+        when :/
+          to_a.my_each { |x| given /= x }
+        end
+        ind
+      else
+        'method error'
+      end
+    else
+      'block not found'
+    end
+  end
+end
```

Comment:
* [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(4) { |prod, n| prod * n } # returns this error
    # NameError:
       undefined local variable or method `length' for 5..50:Range
     # ./lib/main.rb:121:in `my_inject'
  ```

* [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(2, :*) # returns this error
    # NameError:
       undefined local variable or method `ind' for 5..50:Range
       Did you mean?  find
                      end
                      end
                      and
                      END
     # ./lib/main.rb:154:in `my_inject'
  ```